### PR TITLE
SDK-1208. Interfaces to manage CU registration & heartbeats

### PR DESCRIPTION
**Pull request by ms-mega**

### CODE REVIEW

File: src/megaapi_impl.cpp [hunk 1]
```diff
@@ -13266,27 +13296,46 @@ void MegaApiImpl::syncupdate_local_lockretry(bool waiting)
...
 
+    fireOnRequestFinish(request, make_unique<MegaErrorPrivate>(e));
 }
 
 void MegaApiImpl::backupremove_result(const Error& e, handle backupId)
 {
     if (e || backupId != UNDEF)
     {
         LOG_err << "backupremove_result failed: " << MegaError::getErrorString(e);
+        return;
```
Comment (polmr) [inline src/megaapi_impl.cpp:13336]: Request have to end (fireOnRequestFinish) even if they fail

File: include/megaapi.h [hunk 2]
```diff
@@ -3491,6 +3505,38 @@ class MegaRequest
...
+        
+        /**
+         * @brief Returns number of backup files downloaded
+         *
+         * This value is valid for these requests:
+         *  - MegaApi::sendBackupHeartbeat
+         *
+         * @return number of backup files downloaded
+        */
+        virtual int getBackupDownloads();
```
Comment (polmr) [inline include/megaapi.h:3539]: I believe the policy is to try and avoid adding new members to MegaRequest and try to reuse existing ones (that return the same type) and document them in the corresponding getters. @sergiohs84  might have a a saying in this

Comment (ms-mega) [reply include/megaapi.h]: Removed the new members. Using the existing ones.

File: src/megaapi_impl.cpp (hunk 1 above)
Comment (ms-mega) [reply src/megaapi_impl.cpp]: removed the return statement

**Comment by ms-mega** [comment]
trigger compilation

File: src/megaapi_impl.cpp [hunk 3]
```diff
@@ -22752,6 +22829,89 @@ bool MegaApiImpl::tryLockMutexFor(long long time)
...
+
+    if (state > 0)
+    {
+        request->setAccess(state);
+    }
+
+    if (subState > 0)
+    {
+        request->setNumDetails(subState);
+    }
```
Comment (polmr) [inline src/megaapi_impl.cpp:22889]: Those should be allowed to be 0. 
See https://github.com/meganz/sdk/pull/2112#pullrequestreview-483861021

**Review by polmr — DISMISSED** [review]

**Review by sergiohs84 — APPROVED** [review]

**Comment by sergiohs84** [comment]
> megasdk_PR FAILED (3091) -1
Looks like a race in servers: 
```
[ RUN      ] SdkTest.SdkHeartbeatCommands
tests/integration/SdkTest_test.cpp:4679: Failure
Value of: err
  Actual: -9
Expected: MegaError::API_OK
Which is: 0
setBackup failed (error: -9)
[  FAILED  ] SdkTest.SdkHeartbeatCommands (12836 ms)
```

File: src/megaapi_impl.cpp [hunk 4]
```diff
@@ -13367,27 +13370,55 @@ void MegaApiImpl::syncupdate_local_lockretry(bool waiting)
...
 #endif
 
-void MegaApiImpl::backupput_result(const Error&, handle backupId)
+void MegaApiImpl::backupput_result(const Error& e, handle backupId)
 {
+    if (e || backupId == UNDEF)
+    {
+        LOG_err << "backupput_result failed: " << MegaError::getErrorString(e);
+    }
+
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp:13379]: It's actually not needed, all failed (and successful) requests result in a log message with the result: success or error code :)

File: src/megaapi_impl.cpp [hunk 5]
```diff
@@ -6341,7 +6344,7 @@ void MegaApiImpl::loop()
     {
         client->httpio->getMEGADNSservers(&servers, false);
 
-        if (servers.size())
+                    if (servers.size())
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp:6347]:
```suggestion
        if (servers.size())
```

File: src/megaapi_impl.cpp [hunk 6]
```diff
@@ -13367,27 +13370,55 @@ void MegaApiImpl::syncupdate_local_lockretry(bool waiting)
...
+    fireOnRequestFinish(request, make_unique<MegaErrorPrivate>(e));
 }
 
-void MegaApiImpl::backupupdate_result(const Error&, handle /*backupId*/)
+void MegaApiImpl::backupupdate_result(const Error& e, handle backupId)
 {
+    if (e || backupId == UNDEF)
+    {
+        LOG_err << "backupupdate_result failed: " << MegaError::getErrorString(e);
+    }
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp:13393]: Same, not needed

File: src/megaapi_impl.cpp [hunk 7]
```diff
@@ -13367,27 +13370,55 @@ void MegaApiImpl::syncupdate_local_lockretry(bool waiting)
...
 
+    if (requestMap.find(client->restag) == requestMap.end()) return;
+    MegaRequestPrivate* request = requestMap.at(client->restag);
+    if (!request || (request->getType() != MegaRequest::TYPE_BACKUP_PUT)) return;
+    request->setParentHandle(backupId);
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp:13398]: For updates, it's already set as input param. If you want to, we can add an assert, but not really needed to re-assign it.
```suggestion
    assert(backupId == request->getParentHandle()
```

File: src/megaapi_impl.cpp [hunk 8]
```diff
@@ -13367,27 +13370,55 @@ void MegaApiImpl::syncupdate_local_lockretry(bool waiting)
...
+    MegaRequestPrivate* request = requestMap.at(client->restag);
+    if (!request || (request->getType() != MegaRequest::TYPE_BACKUP_PUT_HEART_BEAT)) return;
 
+    fireOnRequestFinish(request, make_unique<MegaErrorPrivate>(e));
 }
 
 void MegaApiImpl::backupremove_result(const Error& e, handle backupId)
 {
-    if (e || backupId != UNDEF)
+    if (e || backupId == UNDEF)
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp:13413]: Again, this block should be removed.

File: src/megaapi_impl.cpp [hunk 9]
```diff
@@ -15635,12 +15666,12 @@ void MegaApiImpl::getpsa_result(error e, int id, string *title, string *text, st
...
-            request->setFile(image->c_str());
-            request->setPassword(buttontext->c_str());
-            request->setLink(buttonlink->c_str());
-        }
+        request->setName(title->c_str());
+        request->setText(text->c_str());
+        request->setFile(image->c_str());
+        request->setPassword(buttontext->c_str());
+        request->setLink(buttonlink->c_str());
+    }
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp:15674]:
```suggestion
        request->setName(title->c_str());
            request->setText(text->c_str());
            request->setFile(image->c_str());
            request->setPassword(buttontext->c_str());
            request->setLink(buttonlink->c_str());
        }
```

File: src/megaapi_impl.cpp [hunk 10]
```diff
@@ -20355,8 +20386,8 @@ void MegaApiImpl::sendPendingRequests()
                         nid = AttrMap::string2nameid("lbl");
                         remove = (value == LBL_UNKNOWN);
                     }
-                    else
-                    {
+                else
+                {
```
Comment (sergiohs84) [inline src/megaapi_impl.cpp:20390]:
```suggestion
                    else
                    {
```

**Review by sergiohs84 — CHANGES_REQUESTED** [review]

File: include/megaapi_impl.h [hunk 11]
```diff
@@ -1229,6 +1229,9 @@ class MegaRequestPrivate : public MegaRequest
 
         MegaBackupListener *getBackupListener() const;
         void setBackupListener(MegaBackupListener *value);
+        void setBackupId(MegaHandle backupId);
+        void setBackupUploads(int ups);
+        void setBackupDownloads(int downs);
```
Comment (sergiohs84) [inline include/megaapi_impl.h:1234]: Unused methods. I'll remove them
```suggestion
```

**Review by sergiohs84 — CHANGES_REQUESTED** [review]

**Review by sergiohs84 — APPROVED** [review]

**Pull request merged**